### ISSUE TITLE
M and N relationshiop

### ISSUE THREAD

**Issue by ashmrz** [ashawkey/torch-ngp#53]
Hi. Thank you very much for you amazing contribution. I wanted to kindly ask about the relationship between M and N ([Here](https://github.com/ashawkey/torch-ngp/blob/979fabb794a5b389513cb48eb7332555a3db80c5/nerf/renderer.py#L273)). How are they related? To be more specific, which entries of rgbs and sigmas are responsible for the calculation of the first pixel in images? 

Thank you very much for your help.

**Comment by ashawkey** [comment]
@ashmrz Hi, to accelerate ray marching, we sample different number of points per ray, let say there are N rays  (pixels), and for ray i we sample r_i points, then M = r_1 + r_2 + ... + r_N.
We record r_i and offset (o_j = \sum_{i=1}^{j-1}r_i) in the variable called `rays` (check [here](https://github.com/ashawkey/torch-ngp/blob/main/raymarching/raymarching.py#L68)), so you can retrieve the points for the i_th ray by `sigmas[rays[i, 1]: rays[i, 1] + rays[i, 2]]`, but to efficiently do this for all rays, you'll have to modify the CUDA kernels.

**Comment by ashmrz** [comment]
@ashawkey Thank you for your in-depth explanation. It all makes sense now.

**Comment by ashmrz** [comment]
@ashawkey As a suggestion, do you think it would be possible for you to implement the support for the depth-supervised loss function from the paper DS-NeRF that uses sparse depth information from Colmap? It would really help.

**Comment by ashawkey** [comment]
Currently I'm trying to implement some other features, but I'll surely consider DS-NeRF later.

**Comment by ashmrz** [comment]
> @ashmrz Hi, to accelerate ray marching, we sample different number of points per ray, let say there are N rays (pixels), and for ray i we sample r_i points, then M = r_1 + r_2 + ... + r_N. We record r_i and offset (o_j = \sum_{i=1}^{j-1}r_i) in the variable called `rays` (check [here](https://github.com/ashawkey/torch-ngp/blob/main/raymarching/raymarching.py#L68)), so you can retrieve the points for the i_th ray by `sigmas[rays[i, 1]: rays[i, 1] + rays[i, 2]]`, but to efficiently do this for all rays, you'll have to modify the CUDA kernels.

Is this a flag to turn off this accelerated ray marching? I have found that when using torch-NGP with clip, since there are almost no samples around the empty portions, the CLIP loss is far less effective in changing the geometry of the object than it is in changing the color of the existing object.

**Comment by ashawkey** [comment]
The CUDA raymarching is controled by `density_grid`, if you want to sample around 'empty' region, you can manually set `density_grid` to a large positive value (so it is viewed as non-empty) and avoid calling `update_extra_states()` (which update density_grid by querying the network). 
However, in my exps the geometry is very hard to be modifed as desired.
You may also check my [dreamfields](https://github.com/ashawkey/dreamfields-torch) implementation if you are interested in CLIP-guided training.
